src/Filters/ExactMatch.php

```diff
@@ -92,6 +102,11 @@ public function getChildren()
     /**
      * Get a list of file paths to exclude.
      *
+     * @deprecated 3.9.0 Overload the `getBlockedFiles()` method instead.
+     *                   The `getBlockedFiles()` method will be made abstract and therefore required
+     *                   in v4.0 and this method will be removed.
+     *                   If both methods are implemented, the new `getBlockedFiles()` method will take precedence.
+     *
      * @return array
      */
     abstract protected function getBlacklist();
```

Comment:
Can we replace the method with a non- abstract implementation, so downstream code won't need to implement a deprecated method?

Reply:
@derrabus Thanks for the feedback.

I was struggling a little with whether to do that or not, as that would mean we can no longer enforce that at least one of each of `getBlacklist/getBlockedFiles` and `getWhitelist/getAllowedFiles` is required to be implemented and making the _new_ methods `abstract` would introduce a breaking change in a minor.

Only way I can think of still safeguarding it, would be to remove all four method declarations from the `ExactMatch` class and using `method_exists()` checks and throwing an exception is any of these sets don't have a corresponding method, but removing the method declarations also makes it much less obvious what the class expects and would require significantly more documentation to be added to the class.

Not sure it's worth the overhead for the (hopefully) short period until the 4.0 release, especially considering that based on a simple code search, there are no external classes extending the `ExactMatch` class.

What do you think ?

Reply:
The abstract class could contain implementations of both methods calling each other, with a recursion detection. If a recursion is detected, you throw because in that case the downstream class hasn't implemented either of them.

Reply:
@derrabus Hmm.. feels a bit like overengineering for something which isn't a real life issue anyway - see the [impact analysis](https://sourcegraph.com/search?q=context:global+lang:php+/extends+ExactMatch%5Cb/+-file:%28%5E%7C%5C/%29vendor%5C/squizlabs%5C/php_codesniffer%5C/&patternType=standard&sm=1&groupBy=repo) (same link I posted in the underlying issue)

The PHP native `implements Serializable` vs `__serialize()` conundrum also comes to mind.

Based on the code in this PR, the following situations are possible for custom `Filter` classes which extend `ExactMatch`:
1. Do nothing, i.e. only implements the old `abstract` methods.
    Impact:
    - Will continue to work fine without deprecation notices in combination with any PHPCS 3.x release.
    - Will break in combination with PHPCS 4.0 or higher.
2. Implement both the old and the new methods in whichever way they prefer (one pointing to the other or visa versa).
    Impact:
    - Will work cross-version with any PHPCS 3.x or 4.x release without notices.
    - Once support for PHPCS 3.x is dropped, the implementation of the "old" methods can be removed.
3. Implement only the "new" methods.
    Impact:
    - Will not work in combination with PHPCS 3.x.
    - Well, that would be one package which is ready early for PHPCS 4.x ;-)